### ISSUE TITLE
Review-2018.09.16-刘文杰

### ISSUE THREAD

**Issue by liuwenjie1995** [DIYer22/weekly-review#287]
本周工作:
1.编写开题报告以及ppt,将文件于9月20提交给李老师进行审核.
2.查看pytorch的文档,学习一些相关API

实验:
1.编写 MNIST模型,观察tensorflow与torch的区别

import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F 
import torch.optim as optim
from torchvision import datasets,transforms
from torch.autograd import Variable

parser = argparse.ArgumentParser(description="this is mnist test for torch")
parser.add_argument('--batch-size','-b',default=64,type=int,metavar = 'BATCHSIZE',help="batch size")
parser.add_argument('-test-batchsize','-t',default=1000,type=int, metavar='N')
parser.add_argument('--epochs','-p' ,default=100,metavar='EPOCH')
parser.add_argument('--lr',default=0.01,metavar='LR',help="training learning rate")
parser.add_argument('--momentum',type=float,default=0.5,metavar='M')
parser.add_argument('--no-cuda',action='store_true',default=False,help='using cuda manager')
parser.add_argument('--seed',type=int,default=1,metavar='S')
parser.add_argument('--log-interval',type=int,default=10,metavar='LOG')

args = parser.parse_args()
args.cuda = not args.no_cuda and torch.cuda.is_available()

torch.manual_seed(args.seed)
if args.cuda:
      torch.cuda.manual_seed(args.seed)

kwargs = {'num_workers':1,'pin_memory':True} if args.cuda else {}

train_loader = torch.utils.data.DataLoader(
      datasets.MNIST('../data',train=True,
      transform=transforms.Compose([
            transforms.ToTensor(),
            transforms.Normalize((0.1307,),(0.3081,))
      ])
      ),
      batch_size=args.batch_size,shuffle=True,**kwargs
)
test_loader = torch.utils.data.DataLoader(
      datasets.MNIST('../data',train=False,transform=
            transforms.Compose([
                  transforms.ToTensor(),
                  transforms.Normalize((0.1307,),(0.3081,))
            ])
      ),
      batch_size=args.batch_size,shuffle=True,**kwargs
)
class Net(nn.Module):
      def __init__(self):
            super(Net,self).__init__()
            self.conv1 = nn.Conv2d(1,10,kernel_size=5)
            self.conv2 = nn.Conv2d(10,20,kernel_size=5)
            self.conv2_drop = nn.Dropout2d()
            self.fc1 = nn.Linear(320,50)
            self.fc2 = nn.Linear(50,10)

      def forward(self,x):
            x = F.relu(F.max_pool2d(self.conv1(x),2))
            x = F.relu(F.max_pool2d(self.conv2_drop(self.conv2(x)),2))
            x = x.view(-1,320)
            x = self.fc1(x)
            x = F.relu(x)
            x = F.dropout(x,training=self.training)
            x = self.fc2(x)
            return F.log_softmax(x)
            

model = Net()

model =torch.nn.Sequential(
      torch.nn.Conv2d(1,10,kernel_size=5),
      torch.nn.MaxPool2d(2),
      torch.nn.Conv2d(10,20,kernel_size=5),
      torch.nn.Dropout(0.5),
      
      torch.nn.Linear(320,50),
      torch.nn.Dropout(0.5),
      torch.nn.Linear(50,10)
)

print(model)

if args.cuda:
      model.cuda()

optimizer = optim.SGD(model.parameters(),lr=args.lr,momentum=args.momentum)

def train(eopch):
      model.train()
      for batch_idx,(data,target) in enumerate(train_loader):
            if args.cuda:
                  data,target = data.cuda(),target.cuda()
            data,target = Variable(data),Variable(target)
            optimizer.zero_grad()
            output = model(data)
            loss = F.nll_loss(output,target)
            loss.backward()
            optimizer.step()
            if batch_idx % args.log_interval == 0 :
                  print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                        epoch, batch_idx * len(data), len(train_loader.dataset),
                        100. * batch_idx / len(train_loader), loss.data[0]))
                        
def test(epoch):
      model.eval()
      test_loss = 0
      correct = 0
      for data,target in test_loader:
            if args.cuda:
                  data,target = data.cuda(),target.cuda()
            data,target = Variable(data,volatile=True),Variable(target)
            output = model(data)
            test_loss += F.nll_loss(output,target).data[0]
            pred = output.data.max(1)[1]
            correct += pred.eq(target.data).cpu().sum()

      test_loss = test_loss
      test_loss /= len(test_loader)
      print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
            test_loss, correct, len(test_loader.dataset),
            100. * correct / len(test_loader.dataset)))

for epoch in range(1,args.epochs + 1):
      train(epoch)
      test(epoch)

收获:
1. tensorflow与pytorch初始化方式不一致,使用tensorflow需要将所有层与变量进行初始化,但是pytorch中Tensor与Variable进行合并,创建一个tensor就是在网络中初始化了,所以简单方便一些.
2. tensorflow中optim需要将loss编写进来,但是pytorch中若loss指在计算梯度,无需将loss交给optim,optim可以自动识别loss进行反向求导并减小梯度.
3. tensorflow中数据流图是将所有的网络层写在一个model里,但是pytorch分成了__init__()和forward()层,init层将保存有variable的层,如卷积层和maxpool层,forward层将init层中的参数进行连接,将relu层,dropout层保存在此,因为这些层不需要保存参数.
4.pytorch使用cuda的方式更为简单,只需要将参数和网络初始化时加入.cuda()即可.

from __future__ import division
import torch.nn as nn
import torch.nn.functional as F
import torch
from numpy.linalg import svd
from numpy.random import normal
from math import sqrt

class UNet(nn.Module):
    def __init__(self,colordim =1):
        super(UNet, self).__init__()
        self.conv1_1 = nn.Conv2d(colordim, 64, 3)  # input of (n,n,1), output of (n-2,n-2,64)
        self.conv1_2 = nn.Conv2d(64, 64, 3)
        self.bn1 = nn.BatchNorm2d(64)
        self.conv2_1 = nn.Conv2d(64, 128, 3)
        self.conv2_2 = nn.Conv2d(128, 128, 3)
        self.bn2 = nn.BatchNorm2d(128)
        self.conv3_1 = nn.Conv2d(128, 256, 3)
        self.conv3_2 = nn.Conv2d(256, 256, 3)
        self.bn3 = nn.BatchNorm2d(256)
        self.conv4_1 = nn.Conv2d(256, 512, 3)
        self.conv4_2 = nn.Conv2d(512, 512, 3)
        self.bn4 = nn.BatchNorm2d(512)
        self.conv5_1 = nn.Conv2d(512, 1024, 3)
        self.conv5_2 = nn.Conv2d(1024, 1024, 3)
        self.upconv5 = nn.Conv2d(1024, 512, 1)
        self.bn5 = nn.BatchNorm2d(512)
        self.bn5_out = nn.BatchNorm2d(1024)
        self.conv6_1 = nn.Conv2d(1024, 512, 3)
        self.conv6_2 = nn.Conv2d(512, 512, 3)
        self.upconv6 = nn.Conv2d(512, 256, 1)
        self.bn6 = nn.BatchNorm2d(256)
        self.bn6_out = nn.BatchNorm2d(512)
        self.conv7_1 = nn.Conv2d(512, 256, 3)
        self.conv7_2 = nn.Conv2d(256, 256, 3)
        self.upconv7 = nn.Conv2d(256, 128, 1)
        self.bn7 = nn.BatchNorm2d(128)
        self.bn7_out = nn.BatchNorm2d(256)
        self.conv8_1 = nn.Conv2d(256, 128, 3)
        self.conv8_2 = nn.Conv2d(128, 128, 3)
        self.upconv8 = nn.Conv2d(128, 64, 1)
        self.bn8 = nn.BatchNorm2d(64)
        self.bn8_out = nn.BatchNorm2d(128)
        self.conv9_1 = nn.Conv2d(128, 64, 3)
        self.conv9_2 = nn.Conv2d(64, 64, 3)
        self.conv9_3 = nn.Conv2d(64, colordim, 1)
        self.bn9 = nn.BatchNorm2d(colordim)
        self.maxpool = nn.MaxPool2d(2, stride=2, return_indices=False, ceil_mode=False)
        self.upsample = nn.UpsamplingBilinear2d(scale_factor=2)
        self._initialize_weights()

    def forward(self, x1):
        x1 = F.relu(self.bn1(self.conv1_2(F.relu(self.conv1_1(x1)))))
        # print('x1 size: %d'%(x1.size(2)))
        x2 = F.relu(self.bn2(self.conv2_2(F.relu(self.conv2_1(self.maxpool(x1))))))
        # print('x2 size: %d'%(x2.size(2)))
        x3 = F.relu(self.bn3(self.conv3_2(F.relu(self.conv3_1(self.maxpool(x2))))))
        # print('x3 size: %d'%(x3.size(2)))
        x4 = F.relu(self.bn4(self.conv4_2(F.relu(self.conv4_1(self.maxpool(x3))))))
        # print('x4 size: %d'%(x4.size(2)))
        xup = F.relu(self.conv5_2(F.relu(self.conv5_1(self.maxpool(x4)))))  # x5
        # print('x5 size: %d'%(xup.size(2)))

        xup = self.bn5(self.upconv5(self.upsample(xup)))  # x6in
        cropidx = (x4.size(2) - xup.size(2)) // 2
        x4 = x4[:, :, cropidx:cropidx + xup.size(2), cropidx:cropidx + xup.size(2)]
        # print('crop1 size: %d, x9 size: %d'%(x4crop.size(2),xup.size(2)))
        xup = self.bn5_out(torch.cat((x4, xup), 1))  # x6 cat x4
        xup = F.relu(self.conv6_2(F.relu(self.conv6_1(xup))))  # x6out

        xup = self.bn6(self.upconv6(self.upsample(xup)))  # x7in
        cropidx = (x3.size(2) - xup.size(2)) // 2
        x3 = x3[:, :, cropidx:cropidx + xup.size(2), cropidx:cropidx + xup.size(2)]
        # print('crop1 size: %d, x9 size: %d'%(x3crop.size(2),xup.size(2)))
        xup = self.bn6_out(torch.cat((x3, xup), 1) ) # x7 cat x3
        xup = F.relu(self.conv7_2(F.relu(self.conv7_1(xup))))  # x7out

        xup = self.bn7(self.upconv7(self.upsample(xup)) ) # x8in
        cropidx = (x2.size(2) - xup.size(2)) // 2
        x2 = x2[:, :, cropidx:cropidx + xup.size(2), cropidx:cropidx + xup.size(2)]
        # print('crop1 size: %d, x9 size: %d'%(x2crop.size(2),xup.size(2)))
        xup = self.bn7_out(torch.cat((x2, xup), 1))  # x8 cat x2
        xup = F.relu(self.conv8_2(F.relu(self.conv8_1(xup))))  # x8out

        xup = self.bn8(self.upconv8(self.upsample(xup)) ) # x9in
        cropidx = (x1.size(2) - xup.size(2)) // 2
        x1 = x1[:, :, cropidx:cropidx + xup.size(2), cropidx:cropidx + xup.size(2)]
        # print('crop1 size: %d, x9 size: %d'%(x1crop.size(2),xup.size(2)))
        xup = self.bn8_out(torch.cat((x1, xup), 1))  # x9 cat x1
        xup = F.relu(self.conv9_3(F.relu(self.conv9_2(F.relu(self.conv9_1(xup))))))  # x9out

        return F.softsign(self.bn9(xup))

    def _initialize_weights(self):
        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                n = m.kernel_size[0] * m.kernel_size[1] * m.out_channels
                m.weight.data.normal_(0, sqrt(2. / n))
                if m.bias is not None:
                    m.bias.data.zero_()
            elif isinstance(m, nn.BatchNorm2d):
                m.weight.data.fill_(1)
                m.bias.data.zero_()

unet = UNet().cuda()

收获:练手

下周工作:
1. 将开题提交给李老师
2. 将多种gan放在自己的网络中进行测试

**Comment by ilydouble** [comment]
两个框架的区别是什么？
